src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs

```diff
@@ -716,6 +730,86 @@ void ConfigureListenOptions(ListenOptions listenOptions)
         await AssertConnectionResult(stream, true);
     }
 
+    [ConditionalFact]
+    public async Task ServerCertificateChainInExtraStore()
+    {
+        var streams = new List<SslStream>();
+        CertHelper.CleanupCertificates();
+        (var clientCertificate, var clientChain) = CertHelper.GenerateCertificates("SslStream_ClinetCertificate_SendsChain", serverCertificate: false);
+
+        using (var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser))
+        {
+            // add chain certificate so we can construct chain since there is no way how to pass intermediates directly.
+            store.Open(OpenFlags.ReadWrite);
+            store.AddRange(clientChain);
+            store.Close();
+        }
+
+        using (var chain = new X509Chain())
+        {
+            chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllFlags;
+            chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+            chain.ChainPolicy.DisableCertificateDownloads = false;
+            var chainStatus = chain.Build(clientCertificate);
+            // Verify we can construct full chain
+            if (chain.ChainElements.Count < clientChain.Count)
+            {
+                throw new InvalidOperationException($"chain cannot be built {chain.ChainElements.Count}");
+            }
+        }
+
+        void ConfigureListenOptions(ListenOptions listenOptions)
+        {
+            listenOptions.UseHttps(new HttpsConnectionAdapterOptions
+            {
+                ServerCertificate = _x509Certificate2,
+                ServerCertificateChain = clientChain,
+                OnAuthenticate = (con, so) =>
+                {
+                    so.ClientCertificateRequired = true;
+                    so.RemoteCertificateValidationCallback = (sender, certificate, chain, sslPolicyErrors) =>
+                    {
+                        foreach (var c in clientChain)
```

Comment:
@halter73 So this test fails, because certificate and chain are null, am I hooking up the Kestrel events incorrectly with OnAuthenticate?  Reference test in runtime get a chain with this callback but they are using SslStreams directly

https://github.com/dotnet/runtime/blob/1f0582e1ee1e6bf05df41aa4cf0d246584c280c8/src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs#L903

Reply:
Never mind, I wasn't passing in the clientCert in the connection, so looks like this test passes, so I guess things are working?  I had to bring in a lot of infrastructure for this test but at least we can now generate certs chains for testing using runtimes test code!